lib/linked_list.rb

```diff
@@ -12,120 +12,244 @@ def initialize(value, next_node = nil)
-    def find_nth_from_end(n)
-      raise NotImplementedError
-    end
-
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    def has_cycle
-      raise NotImplementedError
-    end
-
-
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    def get_first
-      raise NotImplementedError
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+  
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(value)
+    @head = Node.new(value, @head)
+  end
+  
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    return false if @head.nil?
+    current = @head
+    until current.nil?
+      return true if current.data == value
+      current = current.next
+    end
+    return false
+  end
+  
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    return nil if @head.nil?
+    current = @head
+    max = current.data
+    until current.nil?
+      max = current.data if current.data > max
+      current = current.next
+    end
+    return max
+  end
+  
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    return nil if @head.nil?
+    current = @head
+    min = current.data
+    until current.nil?
+      min = current.data if current.data < min
+      current = current.next
+    end
+    return min
+  end
+  
+  
+  # method that returns the length of the singly linked list
+  def length
+    return 0 if @head.nil?
+    current = @head
+    length = 0
+    until current.nil?
+      length += 1
+      current = current.next
+    end
+    return length
+  end
+  
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    current = @head
+    counter = 0
+    until counter == index
+      return nil if current.nil?
+      counter += 1
+      current = current.next
+    end
+    return current.data
+  end
+  
+  # method to print all the values in the linked list
+  def visit
+    current = @head
+    until current.nil?
+      puts current.data
+      length += 1
+      current = current.next
     end
-
-    # method that inserts a given value as a new last node in the linked list
-    def add_last(value)
-      raise NotImplementedError
+  end
+  
+  # method to delete the first node found with specified value
+  def delete(value)
+    return if @head.nil?
+    return nil if @head.next.nil? && @head.data == value

```

Comment:
What if the list has 1 element and that's the value.

```suggestion
    return nil if @head.next.nil? && @head.data != value
```